config.org
#+STARTUP: nohideblocks indent

Here are some additional functions/macros that could help you configure Doom:
- ~load!~ for loading external *.el files relative to this one
- ~use-package~ for configuring packages
- ~with-eval-after-load~ for running code after a package has loaded
- ~add-load-path!~ for adding directories to the `load-path', relative to
  this file. Emacs searches the `load-path' when you load packages with
  `require' or `use-package'.
- ~map!~ for binding new keys

!!! [[~/dotfiles/.config/emacs/lisp/demos.org][doom lisp demos]]

** Variables
#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

(setq
 ;; Some functionality uses this to identify you,
 ;; e.g. GPG configuration, email clients, file templates and snippets.
 user-full-name "earthian"
 org-directory "~/Documents/org/"
 fill-column 80
 display-fill-column-indicator-column 80
 default-input-method "russian-computer"
 browse-url-secondary-browser-function #'browse-url-xdg-open
 browse-url-generic-program #'eww
 trash-directory "~/.local/share/Trash"
 delete-by-moving-to-trash t
 projectile-project-search-path '(("~/git" . 3)))

(add-to-list 'safe-local-variable-values '(sqlformat-command . sqlfluff))
(add-to-list 'safe-local-variable-values '(sqlformat-args . ("--config=/home/earthian/.config/sqlfluff/config.cfg")))

(setq
 completion-ignore-case t
 +zen-text-scale 0.5
 enable-recursive-minibuffers t ;; helps emacs not hang when vertico recursing
 consult-fd-args '("fd" "--color=never --full-path --absolute-path --hidden"
                   "--exclude .git --exclude .stversions"))

(with-eval-after-load 'ssh-deploy
  (setq
   ;; ssh-deploy-async 1
   ;; ssh-deploy-verbose 0
   ssh-deploy-automatically-detect-remote-changes 0
   ssh-deploy-force-on-explicit-save 1)
  (add-to-list 'ssh-deploy-exclude-list "\.org$")
  (add-to-list 'ssh-deploy-exclude-list "\.gitattributes$")
  (add-to-list 'ssh-deploy-exclude-list "\\.claude/")
  (add-to-list 'ssh-deploy-exclude-list "\\.git-crypt/"))

;; don't ever create new workspaces
(with-eval-after-load 'persp-mode
  (setq-default persp-emacsclient-init-frame-behaviour-override -1))

(with-eval-after-load 'lsp-mode
  (set-formatter! 'yamlfmt '("yamlfmt" "-in") :modes '(yaml-mode yaml-ts-mode))
  (setq-hook! 'yaml-mode-hook +format-with '(yamlfmt)))

#+end_src

*** [[id:310c1ee3-2e64-4a4a-b494-53b90b813d7e][Fonts]] (use =doom/describe-char= [[https://github.com/doomemacs/doomemacs/commit/303dd28db808b42a2397c0f4b9fdd71e606026ff][func]]):

'pmsl nerd' - get all installed nerd fonts

#+begin_src emacs-lisp
(setq
 ;; firacode nerd font
 ;; saucecodepro nerd font - copy of source code pro but with better composite glyphs
 ;; hasklug nerd font
 ;; hack nerd font
 doom-font (font-spec :family "jetbrainsmono nf" :size 17) ;; :weight 'light
 ;; for text modes, like Zen, Org or Markdown.
 doom-variable-pitch-font (font-spec :family "daddytimemono nerd font" :size 17)
 ;; Cyrillyc fonts:
 ;; BellotaText-Regular.ttf -> ../.
 ;; Comfortaa-VariableFont_wght.ttf
 ;; PoiretOne-Regular.ttf
 ;; PressStart2P-Regular.ttf
 ;; ShantellSans-VariableFont_BNCE,
 ;; Unbounded-VariableFont_wght.ttf
 ;; Jura-VariableFont_wght.ttf -> ..
 ;; MontserratAlternates-Regular.ttf
 doom-unicode-font (font-spec :family "bellotatext")) ;; Symbola

;; https://www.reddit.com/r/emacs/comments/8tz1r0/how_to_set_font_according_to_languages_that_i/
;; https://fonts.google.com/?script=Thai
;; Set decent default font for Thai and maybe some other langs in future
;; but *only* if in a graphical context.
;; Set the one language last glyphs of which should override others
;; when both charsets cover the same codepoints.
(when (fboundp #'set-fontset-font)
  (set-fontset-font t 'thai "Krub"))
(dolist (item '(("Krub" . 1.6)))
  (add-to-list 'face-font-rescale-alist item))
#+end_src

*** Hlissner [[https://github.com/hlissner/.doom.d][#source]]
#+begin_src emacs-lisp
(setq
 ;; Line numbers are pretty slow all around. The performance boost of
 ;; disabling them outweighs the utility of always keeping them on.
 display-line-numbers-type nil
 doom-scratch-initial-major-mode 'org-mode) ;; or lisp-interaction-mode
#+end_src

** Keybindings
_describe-function map!_

#+begin_src emacs-lisp
(map! :leader :desc "M-x" "x" 'execute-extended-command)

(map! :map doom-leader-file-map
      "x" 'doom/toggle-scratch-buffer
      "o" 'find-file-other-window)

(map! :map doom-leader-buffer-map
      "x" 'doom/toggle-scratch-buffer
      "o" 'sad-true-switch-last-buffer)

(map! :map doom-leader-file-map "l" (cmd! (consult-fd "/")))

(define-key global-map (kbd "C-,") 'embark-act)

;; unbind doom/reload bind cuz it breaks perspective and fucs up workspaces all
;; the time
(map! :map help-map "r r" nil)

;; Workspace binds
(map! :leader :desc "Previous workspace" "-" '+workspace/other)
(map! :map doom-leader-workspace-map
      :desc "Workspace swap right"
      "]" '+workspace/swap-right)
(map! :map doom-leader-workspace-map
      :desc "Workspace swap left"
      "[" '+workspace/swap-left)

;; diff-hl bindings
(map! :map doom-leader-git-map :desc "Git show hunk" "v" 'diff-hl-show-hunk)
(map! :map doom-leader-git-map :desc "Git show hunk" "{" 'diff-hl-show-hunk-previous)
(map! :map doom-leader-git-map :desc "Git show hunk" "}" 'diff-hl-show-hunk-next)

(map! :map doom-leader-map :desc "Cheat-sh" "h h" 'cheat-sh-search-topic)

;; Other packages
(map! :map evil-window-map :desc "Ace window" "w" 'ace-window)
#+end_src

** Startup and Global modes
#+begin_src emacs-lisp
(global-subword-mode)
(global-auto-revert-mode)
#+end_src

** Hooks
#+begin_src emacs-lisp
(add-hook 'markdown-mode-hook 'auto-fill-mode)
(add-hook 'prog-mode-hook 'display-fill-column-indicator-mode)
(add-hook 'pdf-view-mode-hook (lambda () (pdf-view-midnight-minor-mode 1)))

#+end_src

** Custom f-s & advices
#+begin_src emacs-lisp
(defun advice--center-buffer (orig-fun &rest args)
  "Center butter."
  (evil-scroll-line-to-center (line-number-at-pos)))

(advice-add 'evil-ex-search :after #'advice--center-buffer)
(advice-add 'Info-next-reference :after #'advice--center-buffer)
(advice-add 'Info-prev-reference :after #'advice--center-buffer)

(add-to-list 'doom-after-init-hook #'doom-load-session)

(defun sad-true-switch-last-buffer ()
  "Switch to REAL last open buffer (including buffers starting with *)."
  (interactive)
  (let ((previous-place (car (window-prev-buffers))))
    (when previous-place
      (switch-to-buffer (car previous-place))
      (goto-char (car (last previous-place))))))

;; --- vertico debugging
;; (setq debug-on-error t)

;; (defun force-debug (func &rest args)
;;   (condition-case e
;;       (apply func args)
;;     ((debug error) (signal (car e) (cdr e)))))

;; (advice-add #'vertico--exhibit :around #'force-debug)
#+end_src


** Development
*** LSP
#+begin_src emacs-lisp
;; Disable invasive lsp-mode features
;; https://emacs-lsp.github.io/lsp-mode/tutorials/how-to-turn-off/
(setq
 ;; lsp-ui-sideline-enable nil   ; not anymore useful than flycheck
 lsp-enable-suggest-server-download nil
 lsp-ui-doc-enable nil)        ; slow and redundant with K

(map! :map doom-leader-code-map
      ;; another option is xref-find-definition-other-window
      :desc "Find def other window" "O" 'xref-find-definitions-other-window
      :desc "Inlay hints mode" "I" 'lsp-inlay-hints-mode)
#+end_src

*** TODO Golang (setup golangci lint & fix)
#+begin_src emacs-lisp
;; https://github.com/golang/tools/blob/master/gopls/doc/settings.md

(with-eval-after-load 'go-ts-mode
  (setq
   lsp-go-use-gofumpt t)
  ;; (add-hook 'go-ts-mode-hook #'flymake-golangci-load-backend)
  (add-hook 'go-ts-mode-hook #'lsp-deferred)

  (set-formatter! 'parsefix '("parsefix" "-f" filepath) :modes '(go-mode go-ts-mode))
  (set-formatter! 'golines '("golines" "-m" "80" filepath) :modes '(go-mode go-ts-mode))
  (set-formatter! 'goimports '("goimports" filepath) :modes '(go-mode go-ts-mode))
  (setq-hook! 'go-ts-mode-hook +format-with '(goimports golines))

  ;; (defun lsp-go-install-save-hooks ()
  ;;   (add-hook 'before-save-hook #'lsp-format-buffer t t)
  ;;   (add-hook 'before-save-hook #'lsp-organize-imports t t))
  ;; (add-hook 'go-ts-mode-hook #'lsp-go-install-save-hooks)

  (map! :map go-ts-mode-map :localleader :n "z" 'dap-breakpoint-toggle))

(with-eval-after-load 'lsp-mode
  ;; https://github.com/golang/tools/blob/master/gopls/doc/settings.md
  (lsp-register-custom-settings
   '(("gopls.completeUnimported" t t)
     ("gopls.usePlaceholders" t t)
     ("gopls.hints.assignVariableTypes" t t)
     ("gopls.hints.compositeLiteralFields" t t)
     ("gopls.staticcheck" t t))))
;; NOTE: setting passed args here doesn't work for some reason
;; ("golangci-lint.command" ["golangci-lint" "run" "--out-format=json" "--show-stats=false" "--issues-exit-code=0"])))

;; (setq lsp-golangci-lint-run-args '("--out-format=json" "--show-stats=false" "--issues-exit-code=0"))

;; (lsp-register-client
;;  (make-lsp-client :new-connection (lsp-stdio-connection
;;                                    '("golangci-lint-langserver"))
;;                   :major-modes '(go-ts-mode)
;;                   ;; :activation-fn (lsp-activate-on "go")
;;                   :language-id "go"
;;                   :priority 0
;;                   :server-id 'golangci-lint
;;                   :add-on? t
;;                   :library-folders-fn #'lsp-go--library-default-directories
;;                   :initialization-options (lambda ()
;;                                             (gethash "golangci-lint"
;;                                                      (lsp-configuration-section "golangci-lint"))))))
#+end_src

*** SQL
#+begin_src emacs-lisp
(setq sqlformat-command 'pgformatter)
;; Optional additional args
(setq sqlformat-args '("-s2" "-f2" "-w80" "--no-space-function"))
(add-hook 'sql-mode-hook 'sqlformat-on-save-mode)
#+end_src

*** KILL JS
#+begin_src emacs-lisp
(push '(rjsx-mode . typescript-mode) major-mode-remap-alist)
#+end_src
*** Lua
#+begin_src emacs-lisp
(with-eval-after-load 'lsp-mode
  (setq
   lsp-clients-lua-language-server-bin "/usr/bin/lua-language-server"
   lsp-lua-hint-enable t)
  (lsp-register-client
   (make-lsp-client
    :new-connection (lsp-stdio-connection "lua-language-server")
    :major-modes '(lua-mode lua-ts-mode)
    :priority 1
    :server-id 'lua-ls-local)))
#+end_src

*** Shell
#+begin_src emacs-lisp
(with-eval-after-load 'sh-script
  (set-formatter! 'shfmt
    ;; "-ci" for case-indent is the only difference from default formatter
    '("shfmt" "-filename" filepath "-ci" "-ln"
      (cl-case (bound-and-true-p sh-shell) (sh "posix") (t "bash"))
      (when apheleia-formatters-respect-indent-level
        (list "-i"
              (number-to-string
               (cond (indent-tabs-mode 0)
                     ((boundp 'sh-basic-offset) sh-basic-offset) (t 4)))))))
  (add-hook! 'sh-mode-hook
    (setq
     apheleia-formatter 'shfmt
     ;; want 'remote' only for sh files cuz that's how emacs handles root files
     apheleia-remote-algorithm 'local)))
#+end_src

*** TODO Magit & forge
Magit:
#+begin_src emacs-lisp
(with-eval-after-load 'magit
  (map! :map git-commit-mode-map
        "C-c C-g" #'gptel-magit-generate-message))
(setq
 ;; Don't restore the wconf after quitting magit, it's jarring
 magit-inhibit-save-previous-winconf t
 ;; This will enable gravatars when viewing commits.
 ;; The service used by default is Libravatar.
 magit-revision-show-gravatars '("^Author:     " . "^Commit:     "))
#+end_src

*** Docker
#+begin_src emacs-lisp
(setq! docker-command "podman"
       docker-container-shell-file-name "/bin/zsh")
#+end_src
*** QML | Quickshell
#+begin_src emacs-lisp
(with-eval-after-load 'lsp-mode
  (add-to-list 'lsp-language-id-configuration '(qml-mode . "qml"))
  (lsp-register-client
   (make-lsp-client :new-connection (lsp-stdio-connection "qmlls6")
                    :major-modes '(qml-mode)
                    :priority 0
                    :server-id 'qmlls))
  (setq lsp-qml-server-command "qmlls6")
  (add-hook 'qml-mode-hook (lambda ()
                             ;; (setq-local electric-indent-chars '(?\n ?\( ?\) ?{ ?} ?\[ ?\] ?\; ?,))
                             (lsp-deferred))))
#+end_src


** Apps
*** Elfeed
#+begin_src emacs-lisp
(with-eval-after-load 'elfeed
  (setq
   rmh-elfeed-org-files '("~/Documents/tech/rss/elfeed.org")
   elfeed-db-directory "~/Documents/tech/rss/db"
   ;; "@6-month-ago +unread" will show only unread ones
   ;; and thats how you add tags to filter string
   elfeed-search-filter ""))

(map! :leader
      :prefix ("R" . "RSS")
      "o" #'=rss
      "r" #'elfeed-update
      "l" #'elfeed-goodies/toggle-logs
      "e" #'sad/elfeed-org-export-opml
      "u" #'elfeed-show-tag--unread
      "O" #'elfeed-search-browse-url ;; open in external browser
      ;; to force update everything after you changed elfeed.org - unload db,
      ;; remove the elfeed-db-directory and then elfeed-update
      "U" #'elfeed-db-unload)


;; maybe call this func in 'elfeed-new-entry-hook', but for now manually
(defun sad/elfeed-org-export-opml ()
  "Export Org feeds under `rmh-elfeed-org-files' to an OPML file."
  (interactive)
  (let ((opml-body (cl-loop for org-file in rmh-elfeed-org-files
                            concat
                            (with-temp-buffer
                              (insert-file-contents
                               (expand-file-name org-file "~/Documents/tech/rss/"))
                              (rmh-elfeed-org-convert-org-to-opml
                               (current-buffer))))))
    (with-temp-file "~/Documents/tech/rss/elfeed.opml"
      (insert "<?xml version=\"1.0\"?>\n")
      (insert "<opml version=\"1.0\">\n")
      (insert "  <head>\n")
      (insert "    <title>Elfeed-Org Export</title>\n")
      (insert "  </head>\n")
      (insert "  <body>\n")
      (insert opml-body)
      (insert "  </body>\n")
      (insert "</opml>\n"))))
#+end_src

*** Emms
#+begin_src emacs-lisp
(add-to-list 'doom-after-init-hook #'+emms/mpd-start-music-daemon)

(with-eval-after-load 'emms
  (add-to-list 'emms-player-list 'emms-player-mpd)
  (add-to-list 'emms-info-functions 'emms-info-mpd)
  (add-hook 'emms-player-started-hook #'emms-mpris-enable)
  (add-hook 'emms-player-stopped-hook #'emms-mpris-disable)
  (setq!
   emms-source-file-default-directory "~/Music"
   emms-player-mpd-music-directory "~/Music"
   emms-source-playlist-default-format 'm3u
   emms-mode-line-mode nil
   emms-playing-time-display-mode nil
   emms-info-native--max-vorbis-comment-size 150000
   emms-repeat-playlist t
   emms-playlist-mode-center-when-go t
   emms-show-format "♪ %s"
   emms-browser-default-browse-type 'info-album
   ;; new settings
   ;; covers
   emms-browser-covers #'emms-browser-cache-thumbnail-async
   emms-browser-thumbnail-small-size 32
   emms-browser-thumbnail-medium-size 64)

  (map! :map emms-playlist-mode-map
        :n "q" #'quit-window))

;; run the 'emms-cache-set-from-mpd-all' to fill the Emms cache with the
;; contents of your MusicPD database. The music in your MusicPD database should
;; then be accessible via the Emms browser.

(map! :leader
      (:prefix ("l" . "listen")

       ;; Playback
       :desc "Current playlist buffer" "c" #'emms
       :desc "Pause" "x" #'emms-pause
       :desc "Stop" "X" #'emms-stop
       :desc "Next" "n" #'emms-next
       :desc "Previous" "p" #'emms-previous
       :desc "Shuffle" "S" #'emms-shuffle
       :desc "Loop playlist (toggle)" "L" #'emms-toggle-repeat-playlist
       ;; :desc "Loop track (toggle)" "L" #'emms-toggle-repeat-track
       ;; :desc "Browser / open close" "b" #'emms-smart-browse
       ;; :desc "Bury emms buffers" "q" #'emms-browser-bury-buffer

       ;; Daemon / db bindings
       :desc "Start daemon" "s" #'+emms/mpd-start-music-daemon
       :desc "Restart daemon" "r" #'+emms/mpd-restart-music-daemon
       ;; call this manually for the newly added tracks to show up in emms
       :desc "Update db" "u" #'+emms/mpc-update-database ;; gets called on 'start'
       :desc "Update all + cache" "R" #'emms-player-mpd-update-all-reset-cache

       ;; Play ...
       (:prefix ("l" . "Play")
        :desc "directory" "d" #'emms-play-directory
        :desc "directory tree" "D" #'emms-play-directory-tree
        :desc "files matching regex" "f" #'emms-play-find
        :desc "file" "F" #'emms-play-file
        :desc "current dired" "c" #'emms-play-dired
        :desc "url (ie for streaming)" "u" #'emms-play-url)))
#+end_src

*** Eww
#+begin_src emacs-lisp
;; from within eww its just 'o'
(map! :map doom-leader-open-map :desc "Web Wowser" "e" 'eww)

(setq! +lookup-open-url-fn #'eww)

(with-eval-after-load 'eww
  (setq!
   ;; make emacs always use its own browser for opening URL links
   browse-url-browser-function 'eww-browse-url
   eww-bookmarks-directory "~/Documents/tech/"
   eww-history-limit 500
   eww-auto-rename-buffer 'title)

  (map! :map doom-leader-buffer-map "e" #'eww-switch-to-buffer)

  (map! :map eww-text-map "TAB" nil "<backtab>" nil)
  (map! :map eww-textarea-map "TAB" nil "<backtab>" nil)
  (map! :map shr-map "TAB" nil "<backtab>" nil)

  (map! :map eww-mode-map
        ;; don't need all windows closing
        [remap quit-window] #'quit-window
        :nv "C-k" #'shrface-previous-headline
        :nv "C-j" #'shrface-next-headline
        :n "yy" nil ;; copying is way more useful
        :n "y" nil
        :n "<tab>" #'shrface-outline-cycle
        :n "<backtab>" #'shrface-outline-cycle-buffer)

  (map! :map eww-buffers-mode-map
        :n "C-k" #'eww-buffer-show-previous
        :n "C-j" #'eww-buffer-show-next)

  (map! :map eww-mode-map :localleader
        :n "r" #'eww-reload
        :vn "s" #'eww-search-words
        :n "l" #'shrface-links-consult
        :n "L" #'+eww/jump-to-url-on-page
        :n "." #'shrface-headline-consult
        ;; different from 'headline-consult' since uses separate buffer
        :n "h" #'shrface-occur))

(with-eval-after-load 'dash-docs (setq dash-docs-browser-func #'eww))
#+end_src

**** Shrface
#+begin_src emacs-lisp
;; (add-hook 'eww-after-render-hook #'org-indent-mode) ;; FIXME
(add-hook 'eww-mode-hook #'shrface-mode)
(add-hook 'mu4e-view-mode-hook (lambda () (progn)
                                       (require 'eww)
                                       (shrface-mode)))

(with-eval-after-load 'shrface
  ;; (unless shrface-toggle-bullets
  ;;   (shrface-regexp)
  ;;   (setq-local imenu-create-index-function #'shrface-imenu-get-tree)))
  (require 'shr-tag-pre-highlight)
  (add-hook 'outline-view-change-hook 'shrface-outline-visibility-changed)
  (setq!
   shr-max-width 100
   shr-cookie-policy nil
   ;; shr-width 65
   shr-indentation 3
   tab-width 2
   ;; shrface-toggle-bullets nil
   shr-use-fonts nil
   shrface-href-versatile t

   shr-external-rendering-functions (append '((title . eww-tag-title)
                                              (form . eww-tag-form)
                                              (input . eww-tag-input)
                                              (button . eww-form-submit)
                                              (textarea . eww-tag-textarea)
                                              (select . eww-tag-select)
                                              (link . eww-tag-link)
                                              (meta . eww-tag-meta)
                                              ;; (a . eww-tag-a)
                                              (code . shrface-tag-code)
                                              (pre . shr-tag-pre-highlight))
                                            shrface-supported-faces-alist)))
#+end_src

*** Mu4e
#+begin_src emacs-lisp
(load-file "~/.config/doom/mu4e.el")
#+end_src
*** Proced [[file:~/dotfiles/.config/emacs/.local/straight/repos/evil-collection/modes/proced/evil-collection-proced.el][evil-collection-proced.el]]
#+begin_src emacs-lisp
(setq! proced-tree-flag t
       proced-auto-update-flag t
       proced-enable-color-flag t
       proced-format 'medium
       proced-auto-update-interval 1
       proced-format-alist
       '((short user pid tree pcpu pmem start time (args comm))
         (medium user pid tree pcpu pmem vsize rss ttname state start time comm)
         (long
          user euid group pid tree pri nice pcpu pmem vsize rss ttname state
          start time comm)
         (verbose
          user euid group egid pid ppid tree pgrp sess pri nice pcpu pmem state
          thcount vsize rss ttname tpgid minflt majflt cminflt cmajflt start time
          utime stime ctime cutime cstime etime comm)))
#+end_src

*** Telega
#+begin_src emacs-lisp
(map! :map doom-leader-open-map "c" 'telega)

;; telega can't load images without gui cuz size calculation fails, lookup the
;; 'telega' telegram chat, i did ask about that there. So.. call manually.
;; (add-to-list 'doom-after-init-hook (telega 1))

(setq telega-server-libs-prefix "/usr") ;; cuz aur package installs there

(map! :after telega :leader
      :prefix ("z" . "telegram")
      "a" #'telega-account-switch
      "b" #'telega-switch-buffer
      "c" #'telega-chat-with
      "S" #'telega-chatbuf-attach-scheduled
      "f" #'telega-buffer-file-send
      "s" #'telega-saved-messages
      "/" #'telega-chatbuf-filter-search ;; _ key to cancel search
      "t" #'telega
      "z" #'telega-translate-region
      "u" #'telega-switch-unread-chat)

(map! :after telega :map telega-msg-button-map
      "D" #'telega-msg-delete-dwim)

(with-eval-after-load 'telega
  (telega-notifications-mode)
  (telega-patrons-mode nil)
  (telega-active-stories-mode nil)
  (telega-active-locations-mode nil)
  (global-telega-squash-message-mode)
  ;; (auto-fill-mode)

  (require 'telega-dired-dwim)

  ;; https://zevlg.github.io/telega.el/#using-proxy-to-access-telegram
  (when (string-equal (getenv "AT_HOME") "true")
    (add-hook 'telega-before-auth-hook
              (lambda ()
                (telega--addProxy
                    '(:server "127.0.0.1" :port 20170 :enable t
                      :type (:@type "proxyTypeSocks5"))
                    t))))   ; t = enable-p

  (setq
   ;; paths
   telega-database-dir (expand-file-name "~/.local/share/telega")
   telega-directory (expand-file-name "~/.local/share/telega")
   telega-msg-save-dir (expand-file-name "~/Downloads")
   telega-temp-dir (expand-file-name "~/.local/share/telega/temp")
   telega-cache-dir (expand-file-name "~/.local/share/telega/cache")
   telega-voip-logfile (expand-file-name "~/.local/share/telega/telega-voip.log")
   telega-server-logfile (expand-file-name "~/.local/share/telega/telega-server.log")
   ;; debloat

config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!

(setq
  ;; Some functionality uses this to identify you,
  ;; e.g. GPG configuration, email clients, file templates and snippets.
  user-full-name "sad"
  user-mail-address "418@duck.com"

  doom-theme 'doom-opera
  org-directory "~/org/"
  ;; Line numbers are pretty slow all around. The performance boost of
  ;; disabling them outweighs the utility of always keeping them on.
  ;;    (c) hlissner
  display-line-numbers-type nil
  doom-font (font-spec :family "Source Code Pro" :size 16)

  ;; IMO, modern editors have trained a bad habit into us all: a burning need for
  ;; completion all the time -- as we type, as we breathe, as we pray to the
  ;; ancient ones -- but how often do you *really* need that information? I say
  ;; rarely. So opt for manual completion
  ;;    (c) hlissner
  company-idle-delay nil
  )

;; Here are some additional functions/macros that could help you configure Doom:
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys

(add-to-list 'initial-frame-alist '(fullscreen . maximized))

;; Disable invasive lsp-mode features
(setq
  lsp-ui-sideline-enable nil   ; not anymore useful than flycheck
  lsp-ui-doc-enable nil        ; slow and redundant with K
  lsp-enable-symbol-highlighting nil
  )

;; Modules

;; :editor evil
;; Focus new window after splitting
(setq evil-split-window-below t
      evil-vsplit-window-right t)

;;; :tools magit
(setq
  ;; magit-save-repository-buffers nil
  ;; Don't restore the wconf after quitting magit, it's jarring
  magit-inhibit-save-previous-winconf t
  transient-values '((magit-rebase "--autosquash" "--autostash")
                      (magit-pull "--rebase" "--autostash"))
  )

;; :ui doom-dashboard
;; (setq fancy-splash-image (concat doom-private-dir "splash.png"))
;; Hide the menu for as minimalistic a startup screen as possible.
(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)

;; :app everywhere FIXME: throws error when first loading
(after! emacs-everywhere
  ;; Easier to match with a bspwm rule:
  ;;   bspc rule -a 'Emacs:emacs-everywhere' state=floating sticky=on
  (setq emacs-everywhere-frame-name-format "emacs-anywhere")

  ;; The modeline is not useful to me in the popup window. It looks much nicer
  ;; to hide it.
  (remove-hook 'emacs-everywhere-init-hooks #'hide-mode-line-mode)

  ;; Semi-center it over the target window, rather than at the cursor position
  ;; (which could be anywhere).
  (defadvice! center-emacs-everywhere-in-origin-window (frame window-info)
    :override #'emacs-everywhere-set-frame-position
    (cl-destructuring-bind (x y width height)
        (emacs-everywhere-window-geometry window-info)
      (set-frame-position frame
                          (+ x (/ width 2) (- (/ width 2)))
                          (+ y (/ height 2))))))

;; TODO: how to redefine it without last function? so i don't see link for github
;; (setq +doom-dashboard-functions
;;       (doom-dashboard-widget-banner
;;        doom-dashboard-widget-loaded)
;;       )

;; colors for treemacs icons
(after! treemacs
  (treemacs-follow-mode)
  (setq doom-themes-treemacs-theme "doom-colors"
        treemacs-space-between-root-nodes      nil ;; t
        treemacs-width                         32  ;; 35
        ))
;; If you still want the default icons, this works for me
;; (after! (doom-themes treemacs)
;;   (setq doom-themes-treemacs-theme "Default"))

(setq better-jumper-context 'buffer) ;; for now

(map! :leader "gv" 'git-gutter:popup-hunk)

;; TODO: wrap a func below eith defadvice so it insets '~' before and after
(map! :leader "ik" 'edmacro-insert-key)

;; environment --> major
(global-subword-mode)
(global-display-fill-column-indicator-mode)
(global-auto-revert-mode)

(setq-default
 fill-column 80 ;; 70, must be a default value to work
 display-fill-column-indicator-column 80
 )

;; environment -> hooks
(add-hook 'org-mode-hook 'auto-fill-mode)
(add-hook 'markdown-mode-hook 'auto-fill-mode)

;; kbds
(map! :leader "x" 'execute-extended-command)

;; does this even work? and is it even needed?
(set-docsets! 'js2-mode "JavaScript")

;; Enable Gravatars REVIEW does it even works?
;; This will enable gravatars when viewing commits.
;; The service used by default is Libravatar.
(setq magit-revision-show-gravatars '("^Author:     " . "^Commit:     "))

;; pkgs
(setq projectile-project-search-path '("~/git/"))

;; org

;; (add-hook! 'org-mode-hook (company-mode -1))
;; (add-hook! 'org-capture-mode-hook (company-mode -1))

(setq
  ;; org-src-window-setup 'current-window
  org-ellipsis " ▾ "
  ;; org-tags-column -80
  org-hide-emphasis-markers t
  )

;; didn't work out for me cuz https://orgmode.org/manual/Custom-time-format.html
;; (setq org-time-stamp-custom-formats '("<%a %d-%m-%Y>" . "<%a %d-%m-%Y %H:%M>"))
;; (setq org-display-custom-times t)

;; journal setup
(setq
  org-journal-date-prefix "#+TITLE: "
  org-journal-time-prefix "* "
  org-journal-date-format "%a, %d-%m-%Y"
  org-journal-file-format "%Y-%m-%d.org")

(after! org
  ;; org capture templates redefining
  (setq org-capture-templates
    '(("a" "Tea entry" entry
        (file +org-capture-project-notes-file)
        "%[~/git/tea/templates/newEntry]"
        :jump-to-captured t
        :immediate-finish t
        :clock-in t
        :clock-keep t
        :empty-lines 1)
       ("b" "Buffer settings" plain
         (file +org-capture-project-notes-file)
         "%[~/git/tea/templates/bufferSettings]"
         :prepend t
         :immediate-finish t
         :empty-lines-after 1)
       ("t" "Personal todo" entry
         (file +org-capture-todo-file)
         "* [ ] %?\n%i\n%a" :prepend t)
       ("n" "Personal notes" entry
         (file +org-capture-notes-file)
         "* %u %?\n%i\n%a" :prepend t)
       ("j" "Journal" entry
         (file+olp+datetree +org-capture-journal-file)
         "* %U %?\n%i\n%a" :prepend t)
       ("p" "Templates for projects")
       ("pt" "Project-local todo" entry
         (file +org-capture-project-todo-file)
         "* TODO %?\n%i\n%a" :prepend t)
       ("pn" "Project-local notes" entry
         (file +org-capture-project-notes-file)
         "* %U %?\n%i\n%a" :prepend t)
       ("pc" "Project-local changelog" entry
         (file-headline +org-capture-project-changelog-file "Unreleased")
         "* %U %?\n%i\n%a" :prepend t)
       ("o" "Centralized templates for projects")
       ("ot" "Project todo" entry #'+org-capture-central-project-todo-file "* TODO %?\n %i\n %a" :heading "Tasks" :prepend nil)
       ("on" "Project notes" entry #'+org-capture-central-project-notes-file "* %U %?\n %i\n %a" :heading "Notes" :prepend t)
       ("oc" "Project changelog" entry #'+org-capture-central-project-changelog-file "* %U %?\n %i\n %a" :heading "Changelog" :prepend t))
    )
  (set-face-attribute 'org-link nil
                      :weight 'normal
                      :background nil)
  (set-face-attribute 'org-code nil
                      :foreground "#a9a1e1"
                      :background nil)
  (set-face-attribute 'org-date nil
                      :foreground "#5B6268"
                      :background nil)
  (set-face-attribute 'org-level-1 nil
                      :foreground "#dbf6ff"
                      :background nil
                      :height 1.2
                      :weight 'normal)
  (set-face-attribute 'org-level-2 nil
                      :foreground "#fbeffc"
                      :background nil
                      :height 1.0
                      :weight 'normal)
  (set-face-attribute 'org-level-3 nil
                      :foreground "#b5ddfc"
                      :background nil
                      :height 1.0
                      :weight 'normal)
  (set-face-attribute 'org-level-4 nil
                      :foreground "#faccfc"
                      :background nil
                      :height 1.0
                      :weight 'normal)
  (set-face-attribute 'org-level-5 nil
                      :weight 'normal)
  (set-face-attribute 'org-level-6 nil
                      :weight 'normal)
  (set-face-attribute 'org-document-title nil
                      :foreground "SlateGray1"
                      :background nil
                      :height 1.75
                      :weight 'bold))

;; org-mode agenda options
;; TODO: how to do it with evil embrace?
;; now after typing '<el TAB' u will get code block with 'emacs-lisp' src
(after! org (add-to-list 'org-structure-template-alist
             '("el" . "src emacs-lisp\n")))

(setq input-method-history (list "russian-computer")) ;; FIXME still doesn't switch

(setq lsp-vetur-format-default-formatter-html '"prettier")

;; my attempts to make forge work with custom gitlab url...
;; did not suffice elisp knowledge to do that (not all forge functions were working..)
(after! forge
  (push '("gitlab.medpoint24.ru" "gitlab.medpoint24.ru/api/v4"
          "gitlab.medpoint24.ru" forge-gitlab-repository) forge-alist))

;; ==================== DEV ====================

;; === sh ===
(after! sh-script
  (set-company-backend! 'sh-mode
    '(company-shell :with company-yasnippet)))

;; elm
;; (add-to-list 'company-backends 'elm-company)
;; (add-hook 'elm-mode-hook 'elm-format-on-save-mode)

(after! js2-mode
  (add-hook 'js2-mode-hook #'jest-minor-mode)
  (set-company-backend! 'js2-mode 'company-tide 'company-yasnippet))

;; UI

;; Prevents some cases of Emacs flickering
(add-to-list 'default-frame-alist '(inhibit-double-buffering . t))
